external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaStreamingWrite.scala

```diff
@@ -103,20 +103,18 @@ class KafkaStreamDataWriter(
     // Send is asynchronous, but we can't commit until all rows are actually in Kafka.
     // This requires flushing and then checking that no callbacks produced errors.
     // We also check for errors before to fail as soon as possible - the check is cheap.
-    checkForErrors()
-    producer.flush()
-    checkForErrors()
+    close()
     KafkaWriterCommitMessage
   }
 
-  def abort(): Unit = {}
+  def abort(): Unit = {
+    close()
+  }
 
-  def close(): Unit = {
+  private def close(): Unit = {
+    checkForErrors()
+    producer.flush()
     checkForErrors()
-    if (producer != null) {
```

Comment:
I think this null check has significant meaning. Let's consider the following example:
* Task aquires a producer
* `producer.write` called
* `producer.commit` called which decrements `inUseCount`
* Random exception thrown
* `producer.abort` called which decrements `inUseCount` again

Having null check and setting null after released would solve double release.
The same applies to `KafkaWriteTask.scala`.

Reply:
This doesn't seem to be addressed yet. Given we are now dealing reference count, bugs on double release would break the logic. Could you please comment if you think it is not needed, or address it if you think it's valid?

Reply:
In the later comments @gaborgsomogyi, has himself clarified that it is not required, because now we can safely call double release on a producer.